lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -246,45 +246,43 @@ SmallVector<unsigned> getWarpOrder(Attribute layout) {
 
 SmallVector<unsigned> getOrder(Attribute layout) {
   if (auto blockedLayout = dyn_cast<BlockedEncodingAttr>(layout)) {
-    return SmallVector<unsigned>(blockedLayout.getOrder().begin(),
-                                 blockedLayout.getOrder().end());
-  } else if (auto mmaLayout = dyn_cast<MmaEncodingTrait>(layout)) {
+    return llvm::to_vector(blockedLayout.getOrder());
+  }
+  if (auto mmaLayout = dyn_cast<MmaEncodingTrait>(layout)) {
     auto distributedLayout = cast<DistributedEncodingTrait>(layout);
     auto rank = distributedLayout.getWarpsPerCTA().size();
     SmallVector<unsigned> order(rank);
-    for (auto i = 0; i < rank; ++i)
-      order[i] = rank - 1 - i;
+    std::iota(order.rbegin(), order.rend(), 0);
     if (auto mfmaLayout = dyn_cast<AMDMfmaEncodingAttr>(layout)) {
-      if (mfmaLayout.getIsTransposed()) {
-        std::swap(order[rank - 2], order[rank - 1]);
-      }
+      // For transposed MFMA layouts, we swap M and N dimensions, which is
+      // always the first two in order; as we can have an optional batch
+      // dimension following them.
+      if (mfmaLayout.getIsTransposed())
+        std::swap(order[0], order[1]);
```

Comment:
@antiagainst Does `order[0]` refers to the fastest changing dimension of the tensor? If so, does mfma layout always have order[0] = 'N' dim, whether it's transposed or not?

Reply:
I believe line 255 does this.

Reply:
Great question. I'm missing some historical context regarding how the `isTransposed` bit on mfma layout attribute was introduced/used. @binarman can fill that out.

My understanding is that it's just a "contract" for us to be consistent with register/lane distribution for flash attention. And I can see we do such `isTransposed` checks in a few places across the codebase to swap. So we need to be consistent regarding this. The other way is to drop this `isTransposed` and rely on the patterns/passes to recognize and transpose/relayout when necessary; the argument would be that we don't have a special transposed mode in mfma intrinsics. it's kinda normal trade off in compilers regarding how much you want to bake into IR constructs..

Reply:
I think my question is why we need to swap order when mfma layout is transposed. 
The name `isTransposed` is a little misleading here, since nothing is "transposed". We can actually call `mfma.transposed` layout something else, like mfmaV1 layout. 
Then it makes sense to me that the `order` needs to be swapped because we want to use the same linear layout for both mfma and mfmaV1. But if `getOrder` is used somewhere else, and it refers to the storage order of the output tensor, then mfma and mfmaV1 should have the same order.

Reply:
Yeah agreed that `isTransposed` is misleading--it's "transposed" from the perspective that we need to have the first matmul transposed in the attention; not that the layout itself is transposed (nor is that mfma matrix core has a mode of accepting transposed inputs). that's what I meant by saying this is just a "contract" (in the IR construct for transformation purposes) and the other way is dropping this and have patterns/passes handle it.

Reply:
Hi everyone,
`IsTransposed` is a leftover from the times when we did not have clear understanding what is `order` of layout.
There are one-to-one match between `isTransposed` and layout `order`.

This attribute effectively encodes layout order. For 2d tensors it goes like this:
- non transposed [1, 0] aka [M, N]
- transposed [0, 1] aka [N, M]

After dot 3d was introduced, 2 more formats appeared:
- non transposed [2, 1, 0] aka [M, N, B]
- transposed [1, 2, 0] aka [N, M, B]

P.s. I don't think we should drop this attribute, it will complicate various analysis, conversions and make IR less readable. If this attribute is really that confusing, it is better to replace it with plain `order` in my opinion.